src/cargo/ops/registry/auth.rs

```diff
@@ -0,0 +1,229 @@
+            bail!("no upload token found, please run `cargo login` or pass `--token`");
+        }
+        (Some(cli_token), _, _) => cli_token.to_string(),
+        (None, Some(config_token), _) => config_token.to_string(),
+        (None, None, Some(process)) => {
+            let registry_name = registry_name.unwrap_or(CRATES_IO_REGISTRY);
+            run_command(config, process, registry_name, api_url, Action::Get)?.unwrap()
+        }
+    };
+    Ok(token)
+}
+
+/// Saves the given token.
+pub(super) fn login(
+    config: &Config,
+    token: String,
+    credential_process: Option<&(PathBuf, Vec<String>)>,
+    registry_name: Option<&str>,
+    api_url: &str,
+) -> CargoResult<()> {
+    if let Some(process) = credential_process {
+        let registry_name = registry_name.unwrap_or(CRATES_IO_REGISTRY);
+        run_command(
+            config,
+            process,
+            registry_name,
+            api_url,
+            Action::Store(token),
+        )?;
+    } else {
+        config::save_credentials(config, Some(token), registry_name)?;
+    }
+    Ok(())
+}
+
+/// Removes the token for the given registry.
+pub(super) fn logout(
+    config: &Config,
+    credential_process: Option<&(PathBuf, Vec<String>)>,
+    registry_name: Option<&str>,
+    api_url: &str,
+) -> CargoResult<()> {
+    if let Some(process) = credential_process {
+        let registry_name = registry_name.unwrap_or(CRATES_IO_REGISTRY);
+        run_command(config, process, registry_name, api_url, Action::Erase)?;
+    } else {
+        config::save_credentials(config, None, registry_name)?;
+    }
+    Ok(())
+}
+
+fn run_command(
+    config: &Config,
+    process: &(PathBuf, Vec<String>),
+    name: &str,
+    api_url: &str,
+    action: Action,
+) -> CargoResult<Option<String>> {
+    let cred_proc;
+    let (exe, args) = if process.0.to_str().unwrap_or("").starts_with("cargo:") {
+        cred_proc = sysroot_credential(config, process)?;
+        &cred_proc
+    } else {
+        process
+    };
+    if !args.iter().any(|arg| arg.contains("{action}")) {
+        let msg = |which| {
+            format!(
+                "credential process `{}` cannot be used to {}, \
+                 the credential-process configuration value must pass the \
+                 `{{action}}` argument in the config to support this command",
+                exe.display(),
+                which
+            )
+        };
+        match action {
+            Action::Get => {}
+            Action::Store(_) => bail!(msg("log in")),
+            Action::Erase => bail!(msg("log out")),
+        }
+    }
+    let action_str = match action {
+        Action::Get => "get",
+        Action::Store(_) => "store",
+        Action::Erase => "erase",
+    };
+    let args: Vec<_> = args
+        .iter()
+        .map(|arg| {
+            arg.replace("{action}", action_str)
+                .replace("{name}", name)
+                .replace("{api_url}", api_url)
```

Comment:
This is a slight forward-compatibility hazard in that there's no escaping availble to call the process with `{action}` literally as a string somewhere. Not really the end of the world and would be sort of a pain to handle, but figured it was at least worth mentioning.